R/prep.tree.R

```diff
@@ -70,11 +71,71 @@ prep.tree <- function(
         if (is.null(tree.df$label)) tree.df$child else tree.df$label
         );
 
+    if (('node.col' %in% colnames(tree.df))) {
+        tree.df$node.col[is.na(tree.df$node.col)] <- default.node.colour;
+    } else {
+        tree.df$node.col <- default.node.colour;
+        }
+
+    tree.df$border.col <- apply(
+        tree.df,
+        MARGIN = 1,
+        FUN = function(row) {
+            if (is.na(row['border.col'])) row['node.col'] else row['border.col'];
+        }
+    );
+
+    if ('border.type' %in% colnames(tree.df)) {
+        valid.border.types <- c(
+            'blank',
```

Comment:
Does passing on the `border.type` to the `lty` param directly not work? Writing out the `valid.border.type` and min/max numeric seems redundant since the `lty` param inputs are standardized by the lattice package already. I think it would make more sense to specify that it follows lattice convention it in the documentation.

Reply:
I get what you're saying, and I do worry about tying our code too closely to implementation details in lattice. While it would be nice to completely leave this error handling to lattice, there are two big drawbacks I've found:
- The error messages aren't very useful (see below). They don't include any information relevant to the user's input.
```
Error in grid.Call.graphics(C_polygon, x$x, x$y, index) : 
  invalid line type: must be length 2, 4, 6 or 8
```
- A `gpar` object can be initialized with invalid values. It will only throw an error when it attempts to render with `grid.draw`. This means that an error which could be caught immediately will be deferred to the very end of the process (with a lot of unnecessary processing along the way).

I certainly don't disagree with you. It's not great to need to maintain this alongside lattice. (What if lattice changes their standards in a future version?) But, given that this solves the two problems above, it may be worth it. I wonder if there's a better way to achieve the same level of error handling...

Also, whatever solution we end up going with, we should consider adding it to BPG. We should also reapply the same approach to improve error handling in other similar situations.

Reply:
Ahh okay I think I understand it better. My thinking was that people can easily google lattice errors since its a widely used tool, and it prevents us from being too strict on the line types, e.g. avoid editing CEV if lattice replaced 'dashed' with 'dash'.  

But, in conclusion, I have no strong feelings either way and it will be beneficial for people who are not experts in lattice

Reply:
I should probably think of this less as a layer of validation around lattice, and instead the place where we define the valid linetypes for the package. It's possible that we even implement our own linetypes, so it's not directly tied to lattice. We can use similar terminology to lattice, but we don't need to commit to following lattice here. In fact, if lattice changed some terminology (for example `dotted` to `dot`), we could keep our own API the same, and simply convert values to fit the new lattice requirements inside the package function.

Reply:
I see how that would be convenience so that our package remain backward compatible if lattice changed their terminology. I think I'm not entirely sure if we want to prioritize being as separate from existing packages as possible or utilize their setup and not reinvent the wheel too much. Because can we not tie our releases with specific lattice versions to avoid having to update each time lattice does?